cmd/zpool/zpool_iter.c

```diff
@@ -321,41 +322,259 @@ for_each_vdev(zpool_handle_t *zhp, pool_vdev_iter_f func, void *data)
+				if (strcmp(data->cols[k], uniq_cols[i]) == 0) {
+					/* Is the value width larger? */
+					uniq_cols_width[i] =
+					    MAX(uniq_cols_width[i],
+					    strlen(data->lines[k]));
+				}
+			}
+		}
+	}
+
+	vcdl->uniq_cols = uniq_cols;
+	vcdl->uniq_cols_cnt = cnt;
+	vcdl->uniq_cols_width = uniq_cols_width;
+}
+
+
+/*
+ * Process a line of command output
+ *
+ * When running 'zpool iostat|status -c' the lines of output can either be
+ * in the form of:
+ *
+ *	column_name=value
+ *
+ * Or just:
+ *
+ * 	value
+ *
+ * Process the column_name (if any) and value.
+ *
+ * Returns 1 if line was processed, and there are more lines can still be
+ * processed.
+ *
+ * Returns 0 if this was the last line to process, or error.
+ */
+static int
+vdev_process_cmd_output(vdev_cmd_data_t *data, char *line)
+{
+	char *col = NULL;
+	char *val = line;
+	char *equals;
+	char **tmp;
+
+	if (line == NULL)
+		return (0);
+
+	equals = strchr(line, '=');
+
+	if (equals) {
+		/*
+		 * We have a 'column=value' type line.  Split it into the
+		 * column and value strings by turning the '=' into a '\0'.
+		 */
+		*equals = '\0';
+		col = line;
+		val = equals + 1;
+	} else {
+		val = line;
+	}
+
+	/* Do we already have a column by this name?  If so, skip it. */
+	if (col != NULL) {
+		for (int i = 0; i < data->cols_cnt; i++) {
+			if (strcmp(col, data->cols[i]) == 0)
+				return (1); /* Duplicate, skip */
+		}
+	}
+
+	if (val != NULL) {
+		tmp = realloc(data->lines,
+		    (data->lines_cnt + 1) * sizeof (*data->lines));
+
+		if (tmp == NULL)
+			return (0);
+
+		data->lines = tmp;
+		data->lines[data->lines_cnt] = strdup(val);
+		data->lines_cnt++;
+	}
+
+	if (col != NULL) {
+		tmp = realloc(data->cols,
+		    (data->cols_cnt + 1) * sizeof (*data->cols));
+
+		if (tmp == NULL)
+			return (0);
+
+		data->cols = tmp;
+		data->cols[data->cols_cnt] = strdup(col);
+		data->cols_cnt++;
+	}
+
+	if (val != NULL && col == NULL)
+		return (0);
+
+	return (1);
+}
+
+
+/*
+ * Run the cmd and store results in *data.
+ */
+static void
+vdev_run_cmd(vdev_cmd_data_t *data, char *cmd)
 {
-	vdev_cmd_data_t *data = cb_cmd_data;
 	char *pos = NULL;
 	FILE *fp;
 	size_t len = 0;
-	char cmd[_POSIX_ARG_MAX];
-
-	/* Set our VDEV_PATH and VDEV_UPATH env vars and run command */
-	if (snprintf(cmd, sizeof (cmd), "VDEV_PATH=%s && VDEV_UPATH=\"%s\" && "
-	    "VDEV_ENC_SYSFS_PATH=\"%s\" && %s", data->path ? data->path : "",
-	    data->upath ? data->upath : "",
-	    data->vdev_enc_sysfs_path ? data->vdev_enc_sysfs_path : "",
-	    data->cmd) >= sizeof (cmd)) {
-		/* Our string was truncated */
+	char *argv[2] = {cmd, 0};
+	char *env[5] = {"PATH=/bin:/sbin:/usr/bin:/usr/sbin", NULL, NULL, NULL,
+	    NULL};
+	char *line;
+	int link[2];
+	pid_t pid;
+	int status;
+	int rc;
+
+	if (pipe(link) == -1)
 		return;
+
+	if ((pid = fork()) == -1)
+		return; /* Couldn't fork */
+
+	if (pid == 0) {
+		/* Child process */
+		dup2(link[1], STDOUT_FILENO);
+		close(link[0]);
+		close(link[1]);
```

Comment:
The zed contains a [good example](https://github.com/zfsonlinux/zfs/blob/master/cmd/zed/zed_exec.c#L102) of everything which should be handled as part of the fork/exit.  You're going to want to add the `umask`, `dup2`, and `_exit(127)` handling for the child.  On the parent side since it's an interactive command it's not critical to poll on the children for a timeout, but it might be nice functionality to have.